memiavl/db.go

```diff
@@ -380,26 +381,80 @@ func (db *DB) applyChangeSet(name string, changeSet ChangeSet) error {
 		return errReadOnly
 	}
 
-	var updated bool
-	for _, cs := range db.pendingLog.Changesets {
-		if cs.Name == name {
-			cs.Changeset.Pairs = append(cs.Changeset.Pairs, changeSet.Pairs...)
-			updated = true
-			break
+	db.ensurePendingChangesetMap()
+	if existing, ok := db.pendingChangesets[name]; ok {
+		existing.Changeset.Pairs = append(existing.Changeset.Pairs, changeSet.Pairs...)
+	} else {
+		db.insertPendingChangeSet(&NamedChangeSet{Name: name, Changeset: changeSet})
+	}
+
+	return db.MultiTree.ApplyChangeSet(name, changeSet)
+}
+
+func (db *DB) setPendingChangeSets(changeSets []*NamedChangeSet) {
+	db.pendingLog.Changesets = changeSets
+	if len(changeSets) == 0 {
+		db.clearPendingChangesetMap()
+		return
+	}
+
+	db.rebuildPendingChangesetMap(changeSets)
+}
+
+func (db *DB) ensurePendingChangesetMap() {
+	if len(db.pendingLog.Changesets) == 0 {
+		db.clearPendingChangesetMap()
+		return
+	}
+
+	if db.pendingChangesets != nil {
+		return
+	}
+
+	db.rebuildPendingChangesetMap(db.pendingLog.Changesets)
+}
+
+func (db *DB) clearPendingChangesetMap() {
+	if db.pendingChangesets != nil {
+		clear(db.pendingChangesets)
+	}
+}
+
+func (db *DB) rebuildPendingChangesetMap(changeSets []*NamedChangeSet) {
+	m := db.pendingChangesets
+	if m == nil {
+		m = make(map[string]*NamedChangeSet, len(changeSets))
+	} else {
+		clear(m)
+	}
+	for _, cs := range changeSets {
+		if cs != nil {
+			m[cs.Name] = cs
 		}
 	}
+	db.pendingChangesets = m
+}
 
-	if !updated {
-		db.pendingLog.Changesets = append(db.pendingLog.Changesets, &NamedChangeSet{
-			Name:      name,
-			Changeset: changeSet,
-		})
-		sort.SliceStable(db.pendingLog.Changesets, func(i, j int) bool {
-			return db.pendingLog.Changesets[i].Name < db.pendingLog.Changesets[j].Name
-		})
+func (db *DB) insertPendingChangeSet(cs *NamedChangeSet) {
+	if db.pendingChangesets == nil {
+		db.pendingChangesets = make(map[string]*NamedChangeSet)
 	}
+	db.pendingChangesets[cs.Name] = cs
 
-	return db.MultiTree.ApplyChangeSet(name, changeSet)
+	oldLen := len(db.pendingLog.Changesets)
+	if oldLen == 0 || db.pendingLog.Changesets[oldLen-1].Name < cs.Name {
+		db.pendingLog.Changesets = append(db.pendingLog.Changesets, cs)
+		return
+	}
+
+	idx := sort.Search(oldLen, func(i int) bool {
+		return db.pendingLog.Changesets[i].Name >= cs.Name
```

Comment:
```suggestion
		return db.pendingLog.Changesets[i].Name > cs.Name
```

Shouldnt it be this in case of equality?

Reply:
It is still better to consider the equality case and made it correct, otherwise it may break someday